File: swift/Sources/Hekate/HekateProver.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Oumuamua Labs <info@oumuamua.dev>

/// UniFFI-generated. Do not implement by hand.
public protocol HekateProver {
    associatedtype Inputs: Sendable
    associatedtype Output: Sendable
    associatedtype Token: CancellableToken

    static func prove(inputs: Inputs, cancel: Token) throws -> Output
    static func makeCancelToken() -> Token
}

public extension HekateProver {
    /// `priority: .background` for batch/off-screen
    /// work, `.userInitiated` risks the iOS
    /// background-CPU killer. Cancellation
    /// flows via `token.request()`.
    static func prove(
        inputs: Inputs,
        priority: TaskPriority? = nil
    ) async throws -> Output {
        let token = makeCancelToken()

        return try await withTaskCancellationHandler {
            try await Task.detached(priority: priority) {
                try Self.prove(inputs: inputs, cancel: token)
            }.value
        } onCancel: {
            token.request()
        }
    }
}

File: kotlin/lib/src/main/kotlin/dev/oumuamua/hekate/HekateProver.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Oumuamua Labs <info@oumuamua.dev>

package dev.oumuamua.hekate

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.job
import kotlinx.coroutines.withContext
import kotlin.coroutines.cancellation.CancellationException

/** UniFFI-generated. Do not implement by hand. */
public interface HekateProver<Inputs, Output, Token : CancelToken> {
    public fun prove(inputs: Inputs, cancel: Token): Output
    public fun makeCancelToken(): Token
}

/** `dispatcher` defaults to `Default` (CPU pool, sized to cores).
 *  Pass `IO` or a dedicated pool for concurrent / background proving,
 *  `Default` will starve when multiple proofs run at once.
 *  `Job.cancel()` -> `token.request()`; Rust polls the flag. */
public suspend fun <Inputs, Output, Token : CancelToken> HekateProver<Inputs, Output, Token>.prove(
    inputs: Inputs,
    dispatcher: CoroutineDispatcher = Dispatchers.Default,
): Output {
    val token = makeCancelToken()
    return withContext(dispatcher) {
        val handle = coroutineContext.job.invokeOnCompletion { cause ->
            if (cause is CancellationException) {
                token.request()
            }
        }
        
        try {
            prove(inputs, token)
        } finally {
            handle.dispose()
        }
    }
}